File: luogu/kotlin/src/P1449.kt
import java.util.Stack

object P1449 {
	const val SEPARATOR = '.'

	@JvmStatic
	fun main(args: Array<String>) {
		readLine()?.dropLast(1)?.let { formula ->
			Stack<Int>().let { s ->
				var it = 0
				fun current(): Char = formula[it]
				fun next() {
					++ it
				}

				fun readFullNumber(): Int = buildString {
					while (current() != SEPARATOR) {
						append(current())
						next()
					}
				}.toInt()

				while (it != formula.length) {
					when {
						current().isDigit() -> s.push(readFullNumber())
						else -> {
							val f: Int.(Int) -> Int = when (current()) {
								'+' -> Int::plus
								'-' -> Int::minus
								'*' -> Int::times
								'/' -> Int::div

								else -> TODO()
							}

							s.pop().let { a ->
								s.pop().f(a).run(s::push)
							}
						}
					}

					next()
				}

				println(s.peek())
			}
		}
	}
}

File: luogu/haskell/src/P1449.hs
module P1449 where

readFullNumber' :: String -> String -> (Int, String)
readFullNumber' ('.':xs) s = (read s, xs)
readFullNumber' (x:xs) s = readFullNumber' xs (s ++ [x])

readFullNumber :: String -> (Int, String)
readFullNumber = flip readFullNumber'  ""

-- eval :: Stack -> Formula -> Result
eval :: [Int] -> String -> Int
eval (a:s) [] = a
eval (a:b:s) ('+':xs) = eval (b + a : s) xs
eval (a:b:s) ('-':xs) = eval (b - a : s) xs
eval (a:b:s) ('*':xs) = eval (b * a : s) xs
eval (a:b:s) ('/':xs) = eval (b `div` a : s) xs
eval s xs = eval (number : s) str
  where
    fullNumber = readFullNumber xs
    number = fst fullNumber
    str = snd fullNumber
  
  
main :: IO ()
main = do
  s <- getLine

  print $ eval [] (take (length s - 1) s)
